Mathlib/Order/PartialSups.lean

```diff
@@ -3,103 +3,119 @@ Copyright (c) 2021 Kim Morrison. All rights reserved.
-the sequence `f 0`, `f 0 ⊔ f 1`, `f 0 ⊔ f 1 ⊔ f 2`, ... The point of this definition is that
-* it doesn't need a `⨆`, as opposed to `⨆ (i ≤ n), f i` (which also means the wrong thing on
-  `ConditionallyCompleteLattice`s).
-* it doesn't need a `⊥`, as opposed to `(Finset.range (n + 1)).sup f`.
-* it avoids needing to prove that `Finset.range (n + 1)` is nonempty to use `Finset.sup'`.
+For `β` a preorder in which all bounded-above intervals are finite (such as `ℕ`), and `α` a
+`⊔`-semilattice, we define `partialSups : (β → α) → β →o α` by the formula
+`partialSups f n = (Finset.Iic n).sup' ⋯ f`, where the `⋯` denotes a proof that `Finset.Iic n` is
+nonempty. This is a way of spelling `⊔ k ≤ n, f k` which does not require a `α` to have a bottom
+element, and makes sense in conditionally-complete lattices (where indexed suprema over sets are
+badly-behaved).
 
-Equivalence with those definitions is shown by `partialSups_eq_biSup`, `partialSups_eq_sup_range`,
-and `partialSups_eq_sup'_range` respectively.
+Under stronger hypotheses on `α` and `β`, we show that this coincides with other candidate
+definitions, see e.g. `partialSups_eq_biSup`, `partialSups_eq_sup_range`,
+and `partialSups_eq_sup'_range`.
+
+We show this construction gives a Galois insertion between functions `β → α` and monotone functions
+`β →o α`, see `partialSups.gi`.
 
 ## Notes
 
-One might dispute whether this sequence should start at `f 0` or `⊥`. We choose the former because :
+One might dispute whether this sequence should start at `f 0` or `⊥`. We choose the former because:
 * Starting at `⊥` requires... having a bottom element.
 * `fun f n ↦ (Finset.range n).sup f` is already effectively the sequence starting at `⊥`.
 * If we started at `⊥` we wouldn't have the Galois insertion. See `partialSups.gi`.
 
-## TODO
-
-One could generalize `partialSups` to any locally finite bot preorder domain, in place of `ℕ`.
-Necessary for the TODO in the module docstring of `Order.disjointed`.
 -/
 
+open Finset
 
-variable {α : Type*}
+variable {α β : Type*}
 
 section SemilatticeSup
 
 variable [SemilatticeSup α]
 
+section Preorder
+
+variable [Preorder β] [LocallyFiniteOrderBot β]
+
 /-- The monotone sequence whose value at `n` is the supremum of the `f m` where `m ≤ n`. -/
-def partialSups (f : ℕ → α) : ℕ →o α :=
-  ⟨@Nat.rec (fun _ => α) (f 0) fun (n : ℕ) (a : α) => a ⊔ f (n + 1),
-    monotone_nat_of_le_succ fun _ => le_sup_left⟩
+def partialSups (f : β → α) : β →o α where
+  toFun n := (Iic n).sup' nonempty_Iic f
+  monotone' _ _ hmn := sup'_mono f (Iic_subset_Iic.mpr hmn) nonempty_Iic
 
-@[simp]
-theorem partialSups_zero (f : ℕ → α) : partialSups f 0 = f 0 :=
+lemma partialSups_apply (f : β → α) (n : β) :
+    partialSups f n = (Iic n).sup' nonempty_Iic f :=
   rfl
 
-@[simp]
-theorem partialSups_succ (f : ℕ → α) (n : ℕ) :
-    partialSups f (n + 1) = partialSups f n ⊔ f (n + 1) :=
-  rfl
+/-- An induction principle for proving something about `partialSups`. -/
```

Comment:
This arguably isn't an induction principle, as it doesn't work with `induction`; that would be the forwards direction of the `iff`.